tensorflow/contrib/opt/python/training/model_average_optimizer.py

```diff
@@ -0,0 +1,581 @@
+        learning machines by incremental block training with intra-block
+        parallel optimization and blockwise model-update filtering," in
+        Proceedings of ICASSP, 2016.
+      use_Nesterov: means the Nesterov-style momentum update is applied on the
+        block level. The default value is true. This can accelerate training
+        with non-zero block_momentum_rate.
+      block_learning_rate: block_learning_rate is always 1.0 or slightly higher
+        than 1.0
+    """
+    if total_num_replicas is None:
+      total_num_replicas = replicas_to_aggregate
+    logging.info(
+        "ModelAverageV1: replicas_to_aggregate=%s; total_num_replicas=%s",
+        replicas_to_aggregate, total_num_replicas)
+    self._replicas_to_aggregate = replicas_to_aggregate
+    self._block_momentum_rate = block_momentum_rate
+    self._block_learning_rate = block_learning_rate
+    self._ma_intervals = ma_intervals
+    self._use_Nesterov = use_Nesterov
+    self._gradients_applied = False
+    self._total_num_replicas = total_num_replicas
+    self._tokens_per_step = max(total_num_replicas, replicas_to_aggregate)
+    self._global_step = None
+    self._sync_token_queue = None
+
+    # The synchronization op will be executed in a queue runner which should
+    # only be executed by one of the replicas (usually the chief).
+    self._chief_queue_runner = None
+
+    # Remember which accumulator is on which device to set the initial step in
+    # the accumulator to be global step. This list contains list of the
+    # following format: (accumulator, device).
+    self._accumulator_list = []
+    # ma_run_hook should be called before SyncReplicasHook
+    self._ma_run_hook = False
+    # name: string. Name of the global variables and related operation on ps.
+    self._name = 'modelAverage'
+
+    self.add_variables_for_betweenGraph()
+
+  def add_variables_for_betweenGraph(self):
+    # change all variables to local variables
+    for v in variables.global_variables():
+      if v.op.name.find(self._name) != -1:
+        raise AssertionError('%s: cannot use \'%s\' as a substr of any name for'
+                             ' ops or variables when calling an '
+                             'ModelAverageOptimizer.' % (v.op.name, self._name))
+      ops.add_to_collection(ops.GraphKeys.LOCAL_VARIABLES, v)
+    # clear global_variables list
+    ops.get_default_graph().clear_collection(ops.GraphKeys.GLOBAL_VARIABLES)
+    # generate new global variables dependent on trainable variables
+    i = 0
+    for v in variables.trainable_variables():
+      if v.op.name.find(self._name) != -1:
+        raise AssertionError('%s: cannot use \'%s\' as a substr of a name for '
+                             'any ops or variables when calling an '
+                             'ModelAverageOptimizer.' % (v.op.name, self._name))
+      v_g = variable_scope.get_variable(name='%s_g%d' % (self._name, i),
```

Comment:
v_g is the global-variable version of each user-defined trainable variable.  They are supposed to be placed on PS device. v_g is used to store averaged model parameters.